### PR TITLE
Improve options

### LINKED ISSUE #6: Add proxy settings ability on Curl Client?
The main utility of Curl versus file_get_content is the proxy ability and HTTP auth. 

But there is now way of set CURL options without hack or manual low level client manipulation like:

```
$client = new \Buzz\Client\Curl();
curl_setopt($client->getCurl(), CURLOPT_PROXY, 'blablabla');
```

If fact, a setOption method is missing on the Curl Client, and perhaps an array of curl option (CURLOPT_*) could be given as a constructor parameter.

What do you think about that?
I might be able to provide a patch as soon as I can.

### LINKED ISSUE #206: MultiCurl: Allow to set a default callback
I'd like to set a default callback for each request using MultiCurl. Not sure this is the best way but it does the trick and keeps the same interface. I can create a PR which would add something like this (hardcoded to the original MultiCurl - without the inheritance OFC ). Or is there a better way?

https://github.com/vaniocz/VanioBuzzBundle/blob/master/Buzz/MultiCurl.php
Why do I need this?

When using MultiCurl, all listeners' `postSend` methods are invoked before the requests are really performed - the responses are not populated so the `postSend` methods are actually useless.

I solved it this way using a special ListenerChain which is aware of the client. But it needs to either be able to set a default multicurl callback or to override the options being used for each request like in this issue - https://github.com/kriswallsmith/Buzz/issues/168

The best would be both :)

https://github.com/vaniocz/VanioBuzzBundle/blob/master/Buzz/BatchListenerChain.php

### LINKED ISSUE #236: Extract options from AbstractClient to a separate object
We should have a value object that we store options in. 

Options should be ones that cannot be added in a listener. Ie, verity ssl protocols, timeout, etc.

### PR DESCRIPTION
**Pull request by Nyholm**
Use option resolver to validate options. I also introduced a `BuzzClientInterface` with the new parameters. 

In this PR I'll make sure the MultiCurl implements BuzzClientInterface. 

This will fix #236, fix #6 and fix #206. It is also related to #208.

### CODE REVIEW

**Comment by Nyholm** [comment]
TODO: 
- [ ] When validating the options we need to exclude all the integer/curl options. 
- [ ] Update the tests
- [ ] Write change log.

File: composer.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
     "require": {
         "php": "^7.1",
         "psr/http-client": "dev-master",
-        "nyholm/psr7": "^0.3"
+        "nyholm/psr7": "^0.3",
+        "symfony/options-resolver": "^4.0"
```
Comment (stof) [inline composer.json]: can you do `^3.4 || ^4.0` to be compatible with the Symfony 3.4 LTS ? 4.0 does not add any new features compared to 3.4

File: lib/Buzz/Client/AbstractClient.php [hunk 2]
```diff
@@ -1,77 +1,98 @@
...
 
-    public function getVerifyHost()
+    /**
+     * Replace all existing options with new.
+     */
+    public function withOptions(array $options): self
     {
-        return $this->verifyHost;
-    }
+        $clone = clone $this;
```
Comment (stof) [inline lib/Buzz/Client/AbstractClient.php]: The client looks like a service to me, so making it clonable looks weird to me

File: composer.json (hunk 1 above)
Comment (Nyholm) [reply composer.json]: Yes, of course. Thank you!

File: lib/Buzz/Client/AbstractClient.php (hunk 2 above)
Comment (Nyholm) [reply lib/Buzz/Client/AbstractClient.php]: Yeah, you are correct. I should remove "withOptions" and "withAddedOptions". A client should only be configured when we construct the object, right?

**Pull request merged**